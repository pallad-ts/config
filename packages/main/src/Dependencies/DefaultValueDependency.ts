import {Dependency} from "../Dependency";

export class DefaultValueDependency<T> extends Dependency<T> {
    constructor(private dependency: Dependency<T>, private defaultValue: T) {
        super();
    }

    getDescription(): string {
        return `${this.dependency.getDescription()} - default: ${this.defaultValue}`;
    }

    async isAvailable() {
        return true;
    }

    protected async retrieveValue(): Promise<T> {
        const isAvailable = await this.dependency.isAvailable();

        if (isAvailable) {
            return this.dependency.getValue();
        }
        return this.defaultValue;
    }

    static create<T>(dependency: Dependency<T>, defaultValue?: T) {
        if (defaultValue !== undefined) {
            return new DefaultValueDependency(dependency, defaultValue);
        }
        return dependency;
    }
}