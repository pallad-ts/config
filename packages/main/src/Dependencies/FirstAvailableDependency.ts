import {Dependency} from "../Dependency";

export class FirstAvailableDependency<T> extends Dependency<T> {
    private dependencies: Array<Dependency<T>>;
    private cache?: Promise<Dependency<T> | undefined>;

    constructor(...dependencies: Array<Dependency<T>>) {
        super();
        this.dependencies = dependencies;
    }

    async isAvailable(): Promise<boolean> {
        const dep = await this.getFirstAvailableDependency();
        return !!dep;
    }

    private getFirstAvailableDependency() {
        if (this.cache) {
            return this.cache;
        }

        this.cache = (async () => {
            for (const dep of this.dependencies) {
                if (await dep.isAvailable()) {
                    return dep;
                }
            }
        })();
        return this.cache;
    }

    getDescription(): string {
        const desc = this.dependencies.map(x => x.getDescription()).join(', ');
        return `First available - ${desc}`;
    }

    async retrieveValue(): Promise<T> {
        return (await this.getFirstAvailableDependency())!.getValue();
    }
}