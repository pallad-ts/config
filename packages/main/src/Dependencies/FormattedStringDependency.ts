import {Dependency} from "../Dependency";

function produce(strings: readonly string[], values: any[]) {
    let str = '';
    for (const [index, string] of strings.entries()) {
        str += `${string}${(values[index] || '')}`;
    }
    return str;
}

export class FormattedStringDependency extends Dependency<string> {
    constructor(private strings: readonly string[], private values: any[]) {
        super();
    }

    getDescription(): string {
        const desc = produce(this.strings, this.values.map(x => Dependency.is(x) ? `\${${x.getDescription()}}` : x));
        return `Formatted string ("${desc}")`;
    }

    async isAvailable(): Promise<boolean> {
        const results = await Promise.all(
            this.getDependencies()
                .map(x => x.isAvailable())
        );
        return results.every(x => x);
    }

    getDependencies(): Array<Dependency<any>> {
        return this.values.filter(Dependency.is);
    }

    async retrieveValue(): Promise<any> {
        const values = await Promise.all(
            this.values.map(x => Dependency.is(x) ? x.getValue() : x)
        );
        return produce(this.strings, values);
    }

    static create(strings: readonly string[], ...values: any[]) {
        if (values.some(Dependency.is)) {
            return new FormattedStringDependency(strings, values);
        }
        return produce(strings, values);
    }
}
