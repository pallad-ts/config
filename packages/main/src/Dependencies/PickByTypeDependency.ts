import {ConfigError} from '../ConfigError';
import {Dependency} from '../Dependency';
import {getDependencies} from '../getDependencies';
import {Config} from '../Config';
import {load} from '../load';

export class PickByTypeDependency<T> extends Dependency<T> {
    private options = new Map<string, any>();

    private resolvedType?: string;

    private constructor(private typeDependency: Dependency<string>) {
        super();
    }

    static create(typeDependency: Dependency<string>) {
        return new PickByTypeDependency<undefined>(typeDependency);
    }

    registerOptions<TType extends string, TValue>(
        type: TType,
        value: TValue
    ): PickByTypeDependency<NonNullable<T> | PickByTypeDependency.Value<TType, Config.ResolvedValue<TValue>>> {
        if (this.options.has(type)) {
            throw new ConfigError(`Type "${type}" already exists`);
        }
        this.options.set(type, value);
        return this as any;
    }

    private async getType() {
        if (this.resolvedType) {
            return this.resolvedType;
        }
        const type = this.resolvedType = await this.typeDependency.getValue();
        if (!this.options.has(type)) {
            throw new ConfigError(`Type "${type}" is not supported. Supported values: ${Array.from(this.options.keys()).join(', ')}`);
        }
        return type;
    }

    private async getOptions() {
        const type = await this.getType();
        return this.options.get(type)!;
    }

    protected async retrieveValue() {
        return {
            type: await this.getType(),
            options: await load(await this.getOptions())
        };
    }

    getDescription() {
        return `Pick by type from ${this.typeDependency.getDescription()}`;
    }

    private async getDependencies() {
        const options = await this.getOptions();

        const optionsDependencies = Dependency.is(options) ? [options] : getDependencies(options);
        return [
            this.typeDependency,
            ...optionsDependencies
        ];
    }

    async isAvailable() {
        for (const dep of await this.getDependencies()) {
            const isAvailable = await dep.isAvailable();
            if (!isAvailable) {
                return false;
            }
        }
        return true;
    }
}

export namespace PickByTypeDependency {
    export interface Value<TType extends string, TOptions> {
        type: TType;
        options: TOptions;
    }
}
