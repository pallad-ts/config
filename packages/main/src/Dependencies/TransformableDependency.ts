import {Dependency} from "../Dependency";

export abstract class TransformableDependency<TType> extends Dependency<TType> {
    protected constructor(protected transformer: TransformableDependency.Transformer<TType> = TransformableDependency.identity) {
        super();
    }

    protected abstract retrieveValue(): Promise<any>;

    abstract getDescription(): string;

    abstract isAvailable(): Promise<boolean>;

    async getValue(): Promise<TType> {
        return this.transformer(await super.getValue());
    }
}

export namespace TransformableDependency {
    export function identity(x: any) {
        return x;
    }

    export type Transformer<T> = (x: any) => T;
}