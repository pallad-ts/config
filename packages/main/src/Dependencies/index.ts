export * from './DefaultValueDependency';
export * from './ENVDependency';
export * from './FirstAvailableDependency';
export * from './FormattedStringDependency';
export * from './TransformableDependency';