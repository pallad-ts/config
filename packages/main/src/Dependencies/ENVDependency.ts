import {TransformableDependency} from "./TransformableDependency";

export class ENVDependency<T> extends TransformableDependency<T> {
    constructor(private key: string,
                transformer: TransformableDependency.Transformer<T> = TransformableDependency.identity,
                private envs: typeof process['env'] = process.env) {
        super(transformer);
    }

    getDescription(): string {
        return 'ENV: ' + this.key;
    }

    async isAvailable() {
        if (this.key in this.envs) {
            return this.envs[this.key] !== '';
        }
        return false;
    }

    protected async retrieveValue() {
        return this.envs[this.key];
    }
}
