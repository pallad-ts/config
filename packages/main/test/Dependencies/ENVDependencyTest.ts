import {ENVDependency} from "@src/Dependencies/ENVDependency";
import * as sinon from 'sinon';

describe('ENVDependency', () => {
    describe('isAvailable', () => {
        describe('not available', () => {
            it('not defined', () => {
                const dependency = new ENVDependency('key', undefined, {key2: ''});
                return expect(dependency.isAvailable())
                    .resolves
                    .toEqual(false);
            });

            it('empty string', () => {
                const dependency = new ENVDependency('key', undefined, {key: ''});
                return expect(dependency.isAvailable())
                    .resolves
                    .toEqual(false);
            });
        });

        it('available', () => {
            const dependency = new ENVDependency('key', undefined, {key: 'test'});
            return expect(dependency.isAvailable())
                .resolves
                .toEqual(true);
        });
    });

    it('getting description', () => {
        const dependency = new ENVDependency('key');

        expect(dependency.getDescription())
            .toEqual('ENV: key');
    });

    describe('getting value', () => {
        it('not available', () => {
            const dependency = new ENVDependency('key', undefined, {});
            return expect(dependency.getValue())
                .rejects
                .toMatchSnapshot();
        });

        it('transformed', () => {
            const value = 'bar';
            const envValue = 'foo';
            const transformer = sinon.stub()
                .withArgs(envValue)
                .returns(value);

            const dependency = new ENVDependency('key', transformer, {key: envValue});

            return expect(dependency.getValue())
                .resolves
                .toEqual(value);
        });

        it('raw', () => {
            const value = 'bar';
            const dependency = new ENVDependency('key', undefined, {key: value});

            return expect(dependency.getValue())
                .resolves
                .toEqual(value);
        });
    });
});