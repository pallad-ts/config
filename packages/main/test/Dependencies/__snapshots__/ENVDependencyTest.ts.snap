// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ENVDependency getting value not available 1`] = `[Error: Config value not available: ENV: key]`;
